docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,263 @@
+        kind: MeshService
+        name: c
+      default:
+        action: Allow
+```
+
+Would translate into:
+
+```mermaid
+graph TD
+    A-->D
+    B-->D
+    C-->D
+```
+
+Let's consider other `targetRef` with allow types in combination with `top`/`from` level:
+1. `Mesh/Mesh` - this won't help us because that's the default, anyone can talk to anyone and we can't
+
+2. `MeshService/Mesh`  
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-ms-to-mesh
+spec:
+  targetRef:
+    kind: MeshService
+    name: a
+  from:
+    - targetRef:
+        kind: Mesh
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    B-->A
+    C-->A
+    D-->A
+```
+
+We could either define an edge between every service to that service or have a special list that contains allIncomingAllowed.
+Services from that list would always be added as an outgoing edge.
+
+- `Mesh/MeshService`
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-mesh-to-ms
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    A-->B
+    A-->C
+    A-->D
+```
+
+This is simply `A` can talk to anyone, this could be either adding and outgoing edge from this service to every other service or having it on a list called allOutgoing and not bothering to trim the configuration.
+
+3. `MeshService/MeshService` - already talked about, this is the easy case
+
+4. `MeshServiceSubset` 
+
+Unfortunately there is one option here,
+we need to treat `MeshServiceSubset` just like `MeshService`.
+This is because we can't trim individual endpoints because the data is not available at that point, and
+we have to accept the fact that we will miss out on (small) parts of the config that could be trimmed.
+
+5. `MeshSubset`
+
+This one is a bit more tricky because the potential radious is a bit bigger.
+The approach has to be the same as in `MeshServiceSubset`, meaning: treat `MeshSubset` as just `Mesh`.
+
+There is a case that makes sense using `k8s.kuma.io/namespace: xyz` to allow only traffic from a specific namespace,
+so this needs to be documented that it won't improve the performance.
+
+#### Migration
+
+A typical migration to Mesh involves bringing services gradually into the Mesh.
+If the system is small enough, after several iterations, it will be enabled for the whole system (every service).
+
+If the system is too big to do this we need to do extra work in the iterations (defining MeshTrafficPermissions along the way).
+
+1. 0 services are in the Mesh
+2. Bring X services into the Mesh (where X is small enough, so the Mesh can handle it)
+3. Enable MeshTrafficPermission for services brought into the mesh (the performance improves because we trimmed the configs)
+4. Repeat point 2-3 until all services are in the Mesh
```

Comment:
I don't think that's what happens. Having an MTP for the mesh should be a recommended way to go about a migration. Even if you use a very wide MTP (from: Mesh or from: Namespace).
If you do it organically like this you get transparent benefits.

Reply:
I agree, I just made the example small so it's easier to understand and illustrate. I added a paragraph above.